### PR TITLE
[codegen] Fix SIR stat_comp_benchmark model!

### PR DESCRIPTION
**Pull request by seantalts**
no longer based on #156

### CODE REVIEW

**Comment by seantalts** [comment]
(again, still open to comments on this one!)

File: src/stan_math_backend/Expression_gen.ml [hunk 1]
```diff
@@ -34,9 +34,10 @@ let rec stantype_prim_str = function
   | UArray t -> stantype_prim_str t
   | _ -> "double"
 
-let local_scalar ut = function
-  | DataOnly -> stantype_prim_str ut
-  | AutoDiffable -> "local_scalar_t__"
+let local_scalar ut ad =
+  match (ut, ad) with
+  | _, DataOnly | UInt, AutoDiffable -> stantype_prim_str ut
```
Comment (VMatthijs) [inline src/stan_math_backend/Expression_gen.ml]: What about `UArray UInt, AutoDiffable` etc here? Or do we only care about scalars? (Just want to double check.)

File: test/integration/good/code-gen/cpp.expected [hunk 2]
```diff
@@ -38,28 +47,27 @@ class eight_schools_ncp_model : public prob_grad {
...
     sigma = std::vector<double>(J, 0);
     
     pos__ = 1;
     for (size_t sym1__ = 1; sym1__ <= J; ++sym1__) {
-      stan::model::assign(sigma, stan::model::cons_list(stan::model::index_uni(sym1__), stan::model::nil_index_list()), 
-                          context__.vals_r("sigma")[(pos__ - 1)], "assigning variable sigma[(sym1__ - 1)]");
-      stan::model::assign(pos__, stan::model::nil_index_list(), (pos__ + 1), "assigning variable pos__");
+      assign(sigma, cons_list(index_uni(sym1__), nil_index_list()), context__.vals_r("sigma")
+                                                                    [(pos__ - 1)], "assigning variable sigma[(sym1__ - 1)]");
+      assign(pos__, nil_index_list(), (stan::model::deep_copy(pos__) + 1), "assigning variable pos__");
```
Comment (VMatthijs) [inline test/integration/good/code-gen/cpp.expected]: I remember that the old compiler would warn on making deep copies. Is that a runtime warning that the C++ currently also gives or would it be a compile time warning whenever `deep_copy` code is generated?

File: src/stan_math_backend/Expression_gen.ml (hunk 1 above)
Comment (seantalts) [reply src/stan_math_backend/Expression_gen.ml]: Hmm, the way this is used it's intended to be called on a non-UArray, but it's not super obvious. I'll just add the recursive case to be safe.

File: test/integration/good/code-gen/cpp.expected (hunk 2 above)
Comment (seantalts) [reply test/integration/good/code-gen/cpp.expected]: And this is a particularly ugly place for a deepcopy, given that `pos__` is a compiler-generated integer lol. I can fix this particular one by checking the type to see that it's not an int. 

It's not a runtime warning - I think it probably needs to become stanc3-compile-time warning. Let's talk about it at the meeting and see how we should incorporate that.

**Pull request merged**